Op Reduction, Rebased

gselzer opened the pull request:
This PR rebases the work of #32 on top of #57, making it much easier to merge.

Please see #32 for a description of the workings.

gselzer committed:
Enable optional parameters using reduction

This PR introduces the ability to declare optional parameters in an Op's
**functional method** using the Optional annotation. The motivation
behind this implementation was to maintain from ImageJ Ops the ability
to declare optional parameters without the complexity of field
injection.

Once one parameter (or more) has been declared optional for some Op,
scijava ops will create a set of ReducedOpInfos for that Op; if an Op
has n optional parameters, we will then have n+1 OpInfos that delegate
to that Op (n of which being ReducedOpInfos). These ReducedOpInfos can
then be matched, just as any other OpInfo. **We maintain ImageJ Ops'
behavior of matching subsets of optional parameters in a left-to-right
manner**; this means that if there is an OpMethod:

    public Double foo(Double a, @Optional Double b, @Optional Double c) {}

we can match this Op using:
* a
* a and b
* a, b, and c.

We do not allow the matching of a and c, as when b and c are the same
type, passing a Double would provide no indication whether that Double
should be assigned to b or c. To be deterministic we assume that
optional parameters are always left off **right to left**.

Using this paradigm, users can write a class Op with optional parameters
as

    @Plugin(type = Op.class, name = "foo")
    public  class OptionalArgClass implements BiFunction<Double, Double, Double> {
      public Double apply( Double in1, @Optional in2) {
	if (in2 == null) in2 = <reasonable default>;
	...
      }
    }

When this Op is called *without* the optional parameter, the
ReducedOpInfo will use Javassist to create a *wrapper* for
OptionalArgClass, passing null for each optional parameter that is
omitted in the matched ReducedOpInfo. *note that the Op author is then
responsible for null-checking any optional arguments*.

Using the OpBuilder syntax, this Op can then be called using *either*:

    Double d1 = 1.0;
    Double d2 = 2.0;
    Double o = ops.op("foo").input(d1, d2).outType(Double.class).apply();

or:

    Double d1 = 1.0;
    Double o = ops.op("foo").input(d1).outType(Double.class).apply();

Ops written as methods can similarly be written with optional
parameters:

    @OpMethod(names = "bar", type = BiFunction.class)
    public Double barMethod(Double in1, @Optional Double in2) {
      if (in2 == null) in2 = <reasonable default>;
      ...
    }

When writing an Op as a Field, there are two options.
One can write the Op as either an anonymous Class:

    @OpField(names = "foobar")
    public final BiFunction<Double, Double, Double> foobar = new BiFunction<> {
      public Double apply( Double in1, @Optional in2) {
	if (in2 == null) in2 = <reasonable default>;
	...
      }
    }

or, with the help of a specialized interface, using a lambda:

    public interface BiFunctionWithOptional<I1, I2, I3, O> extends Functions.Arity3<I1, I2, I3, O> {
      @Override
      public O apply(I1 in1, I2 in2, @Optional I3 in3);
    }

    @OpField(names = "baz")
    public final BiFunctionWithOptional<Double, Double, Double, Double> baz = (in1, in2, in3) -> {
      if (in3 == null) in3 = <reasonable default>;
      ...
    };

To promote extensibility, each OpInfo is reduced using a InfoReducer
plugin. InfoReducers are designed to be able to reduce OpInfos backed by
one of a set of functional types (for example, a FunctionReducer should
reduce all Functions). For each functional type an InfoReducer can
reduce, it should be able to produce an OpInfo for any number of
optional parameters. So if, for example, an InfoReducer is able to
reduce a Functions.Arity3, it should be able to reduce that Arity3 into
a BiFunction, a Function, and a Producer. OpInfo reduction is performed
at plugin discovery time, and the Javassist wrapper is created when the
Op is matched.

Closes scijava/scijava#14.

Signed-off-by: Curtis Rueden <ctrueden@wisc.edu>

ctrueden commented:
@gselzer I tried to get started on this PR by:
```
gh pr checkout 58
git reset --hard main
git cherry-pick eaf3fe75d938c14326f4f94fe027e97ec33d7ad2
```
because I think the only commit that is actually new here is the tip.

But there are conflicts. I'll take another crack at it next week. Once I resolve the conflicts, my next steps will be:

1. Verify that everything still actually works.
2. Rewrite the commit message to include all the good details written up in #32.
3. Review the actual patch, commenting on things we should follow up on later.
4. Merge it.

ctrueden commented:
I have now completed (1) and (2) above. @hinerm When next you are reviewing Ops PRs, this one is ready for a look!

hinerm commented on scijava/scijava-function/templates/main/java/org/scijava/function/Inplaces.vm:
```diff
@@ -74,16 +78,41 @@ public final class Inplaces {
+	 * @return the arity of {@code c}, or {@code -1} if {@code c} is <b>not</b> a
+	 *         {@code Inplace}.
+	 */
+	public static InplaceInfo arityOf(Class<?> c) {
```
Why does `InplaceInfo#arityOf` return an `InplaceInfo` instead of an `int`, as the other two interfaces do?

hinerm commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/DefaultOpEnvironment.java:
```diff
@@ -657,6 +657,9 @@ private synchronized void initOpDirectory() {
 		List<OpInfoGenerator> generators = infoGenerators();
 		discoverers.stream().flatMap(d -> d.discover(Op.class).stream()).forEach(o -> registerOpsFrom(o, generators));
 		discoverers.stream().flatMap(d -> d.discover(OpCollection.class).stream()).forEach(o -> registerOpsFrom(o, generators));
+		Set<OpInfo> infos = opDirectory.values().stream().flatMap(Collection::stream).map(info -> opsFromObject(info, generators)).flatMap(Collection::stream).collect(
```
you should be able to just do `flatMap(Collection::stream).forEach(info -> addToOpIndex.accept(info, log));` here and skip building the set.

hinerm replied:
Can't do that since we're iterating over the directory that we'll be adding to.

hinerm commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/reduce/FunctionReducer.java:
```diff
@@ -0,0 +1,38 @@
+	@Override
+	public boolean canReduce(OpInfo info) {
+		boolean isFunction = Functions.isFunction(OpUtils.findFunctionalInterface(Types.raw(info.opType())));
+		boolean canReduce = info.declaredHints().containsNone(
```
Why isn't this check in `ComputerReducer#canReduce`? Can the two classes extend an abstract base class to consolidate the routines? It looks like they are identical except the `xxxx.isXxxxx` and the two `arityOf` calls. So create an abstract superclass with unimplemented abstract methods delegating those three pieces?

hinerm commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/reduce/ReducedOpInfo.java:
```diff
@@ -0,0 +1,222 @@
+	}
+
+	@Override public Hints declaredHints() {
+		return srcInfo().declaredHints().plus(BaseOpHints.Reduction.FORBIDDEN);
```
why not cache this, if necessary, in the constructor? Or lazily if you prefer?

hinerm commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/reduce/ReducedOpInfo.java:
```diff
@@ -0,0 +1,222 @@
+	@Override
+	public String implementationName() {
+		// TODO: improve this name
+		return srcInfo.implementationName() + "Reduction" + paramsReduced; 
```
Could a space or dash between parts of the name help here?

hinerm commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/reduce/ReductionUtils.java:
```diff
@@ -0,0 +1,315 @@
+		Method m = InterfaceInference.singularAbstractMethod(fIface);
+		Class<?> srcFIface = OpUtils.findFunctionalInterface(Types.raw(info.srcInfo().opType()));
+		Method srcM = InterfaceInference.singularAbstractMethod(srcFIface);
+//		Class<?> srcFIface = OpUtils.findFunctionalInterface(Types.raw(info.srcInfo().opType()));
```
remove commented out code (in a separate commit if you want to preserve the history)

hinerm commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/struct/OpResizingMemberParser.java:
```diff
@@ -0,0 +1,63 @@
+	 * 
+	 * @param source the {@link RetypingRequest} from which we create the new
+	 *          {@link List} of {@link Member}s
+	 * @return a new {@link Struct} reflecting the simplified arguments / focused
```
We're not actually returning a `Struct` here right?

hinerm commented on scijava/scijava-ops-engine/src/test/java/org/scijava/ops/engine/reduce/BiFunctionWithOptionalReducer.java:
```diff
@@ -0,0 +1,36 @@
+//package org.scijava.ops.engine.reduce;
```
Delete this class?

hinerm commented on scijava/scijava-ops-spi/src/main/java/org/scijava/ops/spi/Optional.java:
```diff
@@ -0,0 +1,46 @@
+ *
+ * is confusing and hard to read. Which parameters are optional in this case? Is
+ * it obvious that {@code in3} is optional just by looking at {@code Impl}? For
+ * this reason, it should be enforced that the annotation is only on one of the
```
I didn't see any negative case tests, where the annotation appears in both interface and implementation - could you add one to show what errors are produced? Or point me at it if I missed it.

hinerm reviewed (changes requested):
Overall this is looking good and the tests are great to prove that everything is working!

ctrueden commented:
@hinerm Thanks so much for digging into this in detail!

Do you think it would make sense for you to take a crack at addressing your own comments, by pushing additional commits? They look relatively minor to me, and you already have your nose in it, whereas @gselzer would need to dust off his brain. Or if you don't feel comfortable making the changes yourself, maybe a pair programming session is in order?

Alternately: up until this point we were merging every PR in the chain "as is" as long as things fundamentally work, with corrections going onto later PRs. I have no strong preference whether we do that here as well; since the changes are minor it is unlikely to seriously hose up later rebases. But noting these changes to do later on top of a further along main branch is also fine.

hinerm commented:
>Do you think it would make sense for you to take a crack at addressing your own comments, by pushing additional commits? They look relatively minor to me, and you already have your nose in it, whereas @gselzer would need to dust off his brain. Or if you don't feel comfortable making the changes yourself, maybe a pair programming session is in order?

yep I can just make the changes @ctrueden

hinerm committed:
Factor out AbstractInfoReducer

Eliminates the copy-and-paste between Function/Computer reducer and
ensures consistency between both.

hinerm committed:
ReducedOpInfo: cache modified hints

Hints are immutable so instead of creating a new set each time they're
requested, let's cache them once.

hinerm committed:
Remove commented out class/code

hinerm committed:
Fix Javadoc reference
